File: ex240.java
class Solution {
    private static boolean binarySearch(int[] m, int target, int lo, int hi) {
        if (lo > hi)
            return false;

        int mid = (lo + hi) / 2;


        if (target > m[mid])
            return binarySearch(m, target, mid + 1, hi);

        else if (target < m[mid])
            return binarySearch(m, target, lo, mid - 1);

        else
            return true;
    }
    
    public boolean searchMatrix(int[][] matrix, int target) {
        boolean result = false;
        
        for (int i = 0; i < matrix.length; i++) {
            if (binarySearch(matrix[i], target, 0, matrix[i].length - 1) == true) {
                result = true;
                break;
            }
        }
        
        return result;
    }
}

File: ex240.cpp
// Search from the bottom-left corner or the top-right corner

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();    if (m == 0)    return false;
        int n = matrix[0].size();
        
        // Search from the bottom-left corner
        int i = m - 1, j = 0;
        while (i >= 0 && j < n) {
            if (matrix[i][j] == target)    return true;
            else if (target < matrix[i][j])    i--;
            else   j++;
        }
        
        return false;
    }
};